File: day:77/generateParentheses.js
const generateParenthesis = function(n) {
    const output = [];
    function generate(s = '', left = 0, right = 0) {
        if(s.length === 2 * n) {
            output.push(s);
            return;
        }
        if(left < n) {
            generate(s + '(', left + 1, right);
        }
        if(right < left) {
            generate(s + ')' , left, right + 1);
        }
    }
    generate();
    return output;
};

console.log(generateParenthesis(3))

File: day:77/generateParentheses.py
class Solution:
    def generate(self, s, left, right, n, output):
        if len(s) == 2 * n:
            output.append(s);
            return;
        
        if left < n:
            self.generate(s + '(', left + 1, right, n, output);
        if right < left:
            self.generate(s + ')' , left, right + 1, n, output);
            
    def generateParenthesis(self, n: int) -> List[str]:
        output = [];
    
        self.generate('', 0 ,0, n, output);
        return output;
